### PR TITLE
Extend support of Tensor Class(symbolic) to support scalars

### LINKED ISSUE #288: Extend support of Tensor Class(symbolic) to support scalars
🐞 **Problem**
Scalars are also zero order tensors. The Tensor class currently supports 1st order tensors and above.

🎯 **Goal**

  - Make `Tensor` support scalars. This issue can also help in #287 .

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

### PR DESCRIPTION
**Pull request by OVSofia**
Fixes #288

### CODE REVIEW

File: src/einsteinpy/tests/test_symbolic/test_tensor.py [hunk 1]
```diff
@@ -72,13 +72,13 @@ def test_Tensor_repr():
...
 
-def test_TypeError1():
-    arr = 0
-    try:
-        obj = Tensor(arr)
-        assert False
-    except TypeError:
-        assert True
+# def test_TypeError1():
+#    arr = 0
```
Comment (ritzvik) [inline src/einsteinpy/tests/test_symbolic/test_tensor.py]: remove these lines altogether

File: src/einsteinpy/symbolic/tensor.py [hunk 2]
```diff
@@ -91,6 +91,8 @@ def __init__(self, arr, config="ll"):
 
         if isinstance(arr, (list, tuple, np.ndarray)):
             self.arr = sympy.Array(arr)
+        elif isinstance(arr, (int, float)):
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/tensor.py]: Scalars can be sympy expressions also. You need to include that case.
```py
from sympy.core.expr import Expr
# blah-blah codes in between
# check through isinstance
```

Comment (ritzvik) [reply src/einsteinpy/symbolic/tensor.py]: no need for an `elif`
you could have done the same thing in the above `isinstance`.
Like this `if isinstance(arr, (list, tuple, np.ndarray, int, float, Expr)):`

File: src/einsteinpy/tests/test_symbolic/test_tensor.py [hunk 3]
```diff
@@ -147,3 +147,12 @@ def test_BaseRelativityTensor_automatic_calculation_of_free_variables():
...
             and (f in functions)
         )
+
+
+# tests fot Tensor Class to support scalars
+
+
+def test_tensor_scalar():
+    scalar = 11.89
+    scalar_tensor = Tensor(scalar)
```
Comment (ritzvik) [inline src/einsteinpy/tests/test_symbolic/test_tensor.py]: you would need to test for sympy expression type scalars also
```py
import pytest
from sympy.abc import y, z

# all code in between xD

@pytest.mark.parametrize("scalar",[11.89, y*z+5])
def test_tensor_scalar(scalar):
    scalar_tensor = Tensor(scalar)
    assert scalar_tensor.tensor().rank() == 0
```

BTW, it's a nice thing to learn. See this https://docs.pytest.org/en/latest/parametrize.html
pytest parametrize tests more than one input values for a single test function!!

**Comment by ritzvik** [comment]
Also can you edit `RicciScalar` to reflect the new changes?

File: src/einsteinpy/tests/test_symbolic/test_tensor.py (hunk 3 above)
Comment (OVSofia) [reply src/einsteinpy/tests/test_symbolic/test_tensor.py]: Thanks!!

**Comment by OVSofia** [comment]
> Also can you edit `RicciScalar` to reflect the new changes?

Do you want to change code or just docstring?

**Comment by ritzvik** [comment]
> > Also can you edit `RicciScalar` to reflect the new changes?
> 
> Do you want to change code or just docstring?

the code, I was implying. Make it a derived class of `BaseRelativityTensor`

File: src/einsteinpy/symbolic/ricci.py [hunk 4]
```diff
@@ -141,21 +141,24 @@ def change_config(self, newconfig="ul", metric=None):
...
 
-class RicciScalar:
+class RicciScalar(BaseRelativityTensor):
     """
+    Inherits from ~einsteinpy.symbolic.tensor.BaseRelativityTensor
     Class for defining Ricci Scalar
     """
 
-    def __init__(self, expression, syms, parent_metric=None):
+    def __init__(self, arr, syms, config="l", parent_metric=None):
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]: remove the `config` parameter.
`config` is going to be an empty string given scalar is a zero order tensor.

File: src/einsteinpy/symbolic/ricci.py [hunk 5]
```diff
@@ -166,9 +169,10 @@ def __init__(self, expression, syms, parent_metric=None):
             Raised when syms is not a list or tuple
 
         """
+        super(RicciScalar, self).__init__(
+            arr=arr, syms=syms, config=config, parent_metric=parent_metric
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]:
```
super(RicciScalar, self).__init__(
    arr=arr, syms=syms, config="", parent_metric=parent_metric
    )
```

File: src/einsteinpy/symbolic/ricci.py [hunk 6]
```diff
@@ -166,9 +169,10 @@ def __init__(self, expression, syms, parent_metric=None):
...
+        )
         self._order = 0
-        self._parent_metric = parent_metric
-        self._expr = expression
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]: don't remove this line. Change this to `self._expr = self.arr[0]`

File: src/einsteinpy/symbolic/ricci.py [hunk 7]
```diff
@@ -166,9 +169,10 @@ def __init__(self, expression, syms, parent_metric=None):
...
         if isinstance(syms, (list, tuple)):
             self.syms = syms
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]: remove the whole `if isinstance` code block. This is being taken care by `BaseRelativityTensor`

File: src/einsteinpy/symbolic/ricci.py [hunk 8]
```diff
@@ -180,7 +184,7 @@ def expr(self):
         """
         Retuns the symbolic expression of the Ricci Scalar
         """
-        return self._expr
+        return self.arr
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]: leave this unchanged!

File: src/einsteinpy/symbolic/ricci.py [hunk 9]
```diff
@@ -180,7 +184,7 @@ def expr(self):
...
 
     @property
     def parent_metric(self):
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]: remove this property. Already taken care by `BaseRelativityTensor`

File: src/einsteinpy/symbolic/ricci.py [hunk 10]
```diff
@@ -170,28 +168,17 @@ def __init__(self, arr, syms, config="l", parent_metric=None):
 
         """
         super(RicciScalar, self).__init__(
-            arr=arr, syms=syms, config=config, parent_metric=parent_metric
+            arr=arr, syms=syms, config=" ", parent_metric=parent_metric
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/ricci.py]: It was supposed to be a blank string `config=""`

Comment (OVSofia) [reply src/einsteinpy/symbolic/ricci.py]: So silly mistake xD I was looking for what could give so many mistakes in the test

Comment (ritzvik) [reply src/einsteinpy/symbolic/ricci.py]: Well, it was practical lesson why testing is important 😄

Comment (ritzvik) [reply src/einsteinpy/symbolic/ricci.py]: resolve

**Comment by ritzvik** [comment]
There was no need for changing tests in `RicciScalar`, we are not breaking any APIs.

Anyways, pull the new commits, squash and rebase

**Comment by OVSofia** [comment]
> There was no need for changing tests in `RicciScalar`, we are not breaking any APIs.
> 
> Anyways, pull the new commits, squash and rebase


Yes, in fact I was just making the changes of your last commits

File: src/einsteinpy/tests/test_symbolic/test_ricci.py [hunk 11]
```diff
@@ -85,13 +85,13 @@ def test_RicciScalar_is_constant_for_ADS_spacetime():
...
-    assert sympy.simplify(R.arr).is_constant()
+    assert sympy.simplify(R._expr).is_constant()
 
 
-def test_RicciScalar_arr_property():
+def test_RicciScalar_expr_property():
     x, y = sympy.symbols("x y")
     R = RicciScalar(x ** 2 * y * 3, (x, y))
-    assert sympy.simplify(R.arr - R.arr) == 0
+    assert sympy.simplify(R._expr - R._expr) == 0
```
Comment (ritzvik) [inline src/einsteinpy/tests/test_symbolic/test_ricci.py]: `R._expr - R.expr` 
We are testing that what the property returns is same as the value given be attribute.

**Comment by ritzvik** [comment]
you might need to rebase one more time. Sorry xD

**Comment by OVSofia** [comment]
sorry, I had a big mess with git XDD

**Comment by ritzvik** [comment]
> you might need to rebase one more time. Sorry xD

.


> sorry, I had a big mess with git XDD

What happened? 🤣

**Comment by OVSofia** [comment]
> > you might need to rebase one more time. Sorry xD
> 
> .
> 
> > sorry, I had a big mess with git XDD
> 
> What happened?


I had not pushed the last commit because I had seen your changes and started to give conflicts, 
which I have solved in an inefficient way xDD

**Comment by OVSofia** [comment]
@ritzvik I can not see why it's failing :/

**Comment by ritzvik** [comment]
Well the docs started failing out of the blue. I raised the issue in #324

**Comment by ritzvik** [comment]
> Well the docs started failing out of the blue. I raised the issue in #324

@shreyasbapat

**Comment by shreyasbapat** [comment]
Aye Aye!

**Comment by shreyasbapat** [comment]
I hope the dead documentation team was very much useful for you ! (If the test pass at all xD)

**Comment by shreyasbapat** [comment]
They don't!

**Comment by ritzvik** [comment]
> They don't!

The matter of intrigue is how the random code block is popping up?

**Comment by shreyasbapat** [comment]
Who said a thing about a code block?

**Comment by ritzvik** [comment]
> Who said a thing about a code block?

I am talking about this


http://docs.einsteinpy.org/en/latest/api/symbolic/constants.html

**Comment by shreyasbapat** [comment]
WHAT!!!!

**Comment by ritzvik** [comment]
@shreyasbapat 
Sphinx was trying to inherit the docstring of the `__new__` method of the parent class. Hopefully, my commit would resolve it!

**Comment by shreyasbapat** [comment]
Oh! I had no idea about that!

**Comment by ritzvik** [comment]
@shreyasbapat MERGE

**Pull request merged**